src/ParallelAlgorithms/Amr/Criteria/DriveToTarget.hpp

```diff
@@ -0,0 +1,124 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <pup.h>
+
+#include "Domain/Amr/Flag.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace amr::Criteria {
+/*!
+ * \brief Refine the grid towards the target resolution and refinement levels in
+ * each dimension and then oscillate about the target.
+ *
+ * \details If the grid is at neither target in a given dimension, the
+ * flag chosen will be in the priority order Split,
+ * IncreaseResolution, DecreaseResolution, Join.
+ *
+ * \note To remain at the target, set the OscillationAtTarget Flags to
+ * DoNothing.
+ *
+ * \note This criterion is primarily for testing the mechanics of refinement.
+ */
+template <size_t Dim>
+class DriveToTarget : public Criterion {
+ public:
+  /// The target resolution
+  struct TargetResolution {
+    using type = std::array<size_t, Dim>;
+    static constexpr Options::String help = {
+        "The target resolution in each dimension."};
+  };
+
+  /// The target resolution
+  struct TargetRefinementLevels {
+    using type = std::array<size_t, Dim>;
+    static constexpr Options::String help = {
+        "The target refinement levels in each dimension."};
+  };
+
+  /// The AMR flags chosen when the target resolution and refinement levels are
+  /// reached
+  struct OscillationAtTarget {
+    using type = std::array<domain::Flag, Dim>;
+    static constexpr Options::String help = {
+        "The flags returned when at the target."};
+  };
+
+  using options =
+      tmpl::list<TargetResolution, TargetRefinementLevels, OscillationAtTarget>;
+
+  static constexpr Options::String help = {
+      "Refine the grid towards the TargetResolution and "
+      "TargetRefinementLevels, and then oscillate about them by applying "
+      "OscillationAtTarget."};
+
+  DriveToTarget() = default;
+
+  DriveToTarget(const std::array<size_t, Dim>& target_resolution,
+                const std::array<size_t, Dim>& target_refinement_levels,
+                const std::array<domain::Flag, Dim>& flags_at_target);
+
+  /// \cond
+  explicit DriveToTarget(CkMigrateMessage* msg);
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(DriveToTarget);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<::domain::Tags::Mesh<Dim>>;
+
+  template <typename ArrayIndex, typename Metavariables>
+  auto operator()(const Mesh<Dim>& current_mesh,
+                  Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ArrayIndex& array_index) const;
```

Comment:
You rely on ArrayIndex to be an ElementId, so specialize to that and remove the template param I think.

Then you could move the implementation to the cpp (with a private impl function that doesn't depend on the unused metavars). Up to you.